test: Use qemu user and socket based networking

stefwalter opened the pull request:
The test/ code no longer should have anything to do with preparing images for testing. The bots/ code and ```bots/image-prepare``` in particular will take care of this.

 * On master the code in ```test/``` for preparing images will be removed.
 * On branches the code in ```test/``` for preparing images will be unused.

No changes should be required to branches. Fingers crossed.

 * [x] #6672 
 * [x] #6673 
 * [x] #6676 
 * [x] fix changing network names after reboot, #6923 
 * [x] rebuild openshift
 * [x] image-refresh ipa
 * [x] #7060
 * [x] #7105 
 * [x] #7180 
 * [x] fix `check-networking-checkpoints TestNetworking.testCheckpoint` (pitti)
 * [x] fix `check-ostree` reboot failures (stef)

Recreate to get rid of runcmd script:

 * [x] image-refresh rhel-atomic
 * [x] image-refresh fedora-atomic

martinpitt commented on test/images/scripts/openshift.setup:
```diff
@@ -60,18 +61,13 @@ docker cp $container:/usr/bin - | tar -C /usr -xf - bin/openshift bin/oc bin/oad
+/usr/bin/systemctl enable rpcbind
+/usr/bin/systemctl start rpcbind
+/usr/bin/systemctl start nfs-server
+cmd="/usr/bin/openshift start --master=10.111.112.101"
```
AFAIK the "else" branch was for adding a secondary openshift machine for providing additional nodes. This commit would remove that possibility, as the host name is now always fixed to be "f1.cockpit.lan" and it always starts a master. I suppose this will break some of our openshift tests?

stefwalter replied:
No the ability to start an openshift-node was just a nicety. I couldn't figure out a way to make it possible with this setup.

martinpitt commented on test/common/testlib.py:
```diff
@@ -490,38 +497,45 @@ def run(self, result=None):
-    def setUp(self, macaddr=None, memory_mb=None, cpus=None):
+    def setUp(self):
+        if opts.address and self.provision is not None:
+            raise unittest.SkipTest("Cannot provision multiple machines if a specific machine address is specified")
```
IMHO that would just silently paper over errors in our tests. This should rather be a `ValueError` or other fatal exception.

stefwalter replied:
This is identical to current behavior. Certain tests cannot be run with an ```--address``` argument.

stefwalter replied:
If this behavior is to be changed it should be done as a follow up. It only affects running of tests manually against a running system.

martinpitt replied:
Ah, this is only for the CLI option. I see, ignore me then.

martinpitt commented on test/common/testvm.py:
```diff
@@ -96,23 +91,50 @@ def __str__(self):
+            (self.web_address, unused, self.web_port) = browser.rpartition(":")
+        else:
+            self.web_address = address
+            self.web_port = browser or 9090
```
That looks a bit odd. Of what type is "browser"? `new_browser()` returns a `Browser` object, but by the looks of this code it's expected to be a simple value. But the `if` part expects a "hostname:port" string, and the `else` part expects a single number. I don't actually see any instantiation of `Machine` with a non-default `browser` argument, but I think this should rather be removed entirely then.

stefwalter replied:
It's there see VirtNetwork and VirtMachine. I've updated it to make it consistent.

martinpitt commented on test/common/testvm.py:
```diff
@@ -847,57 +858,182 @@ def virEventLoopNativeRun():
+
+        # Currently all images are x86_64. When that changes we will have
+        # an override file for those images that are not
+        self.arch = "x86_64"
```
This is misleading -- we do have an "i386" image now. It can of course run on a 64 bit (virtual) machine. Maybe for now change the comment to "all images run on x86_64"?

stefwalter replied:
This comment was not added merely moved. However I've updated the comment in question.

martinpitt commented on test/images/scripts/fedora-25.setup:
```diff
@@ -76,6 +76,12 @@ maybe() { if type "$1" >/dev/null 2>&1; then "$@"; fi; }
+# If the main VM run a dnsmasq instance
+for x in $(seq 1 9); do
+    nmcli connection add type ethernet con-name machine1 ifname "*" mac ee:00:0a:00:00:$x ip4 10.0.0.$x/24
+done
```
This just adds client-side interfaces in NM, not a DNS server. Fedora doesn't default to using a client-side dnsmasq local DNS resolver. The comment could be clarified, also why it needs a dnsmasq?

Depending on what the purpose is, I figure this needs to be replicated to fedora-{26,testing}.setup? What about debians, Atomics, etc?

stefwalter replied:
This was a bogus change from refactoring. Removed this change.

martinpitt commented on test/verify/check-dashboard:
```diff
@@ -94,100 +92,100 @@ class TestBasicDashboard(MachineCase, DashBoardHelpers):
-        self.wait_dashboard_addresses (b2, [ "localhost", m2.address ])
+        self.add_machine (b, "10.111.113.2")
+        self.wait_dashboard_addresses (b, [ "localhost", "10.111.113.2" ])
+        self.wait_dashboard_addresses (b2, [ "localhost", "10.111.113.2" ])
```
I still find `m2.address` much more robust, readable, and maintainable (IPv6 or changing networking setup again) than hardcoded and massively duplicated IP numbers. Changing the definition above like `m2 = provision["machine2"]` and using `m2["address"]` perhaps?

stefwalter replied:
There were 10's if not 100 interpolation statements with addresses. These are all gone now, which on balance makes things much more readable and maintainable. The fixed IP addresses are now something we can rely on in tests. This is now similar to hard coding things like component names in URLs or HTML CSS selectors such as ```#id```.

martinpitt commented:
There seems to be a race condition with the VM setup, as shown by the debian-8 and rhel-7 failures?

dperpeet commented:
Martin already mentioned this race:

```
ssh_exchange_identification: read: Connection reset by peer
Domain not found: no domain with matching uuid '5f174ecd-074a-4b00-a65e-30bff002f250' (rhel-7-127.0.0.2-2201)
Traceback (most recent call last):
  File "./verify/run-tests", line 39, in install
    subprocess.check_call(cmd)
  File "/usr/lib64/python2.7/subprocess.py", line 186, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['/build/cockpit/test/verify/testsuite-prepare']' returned non-zero exit status 2
Install failed
```

stefwalter commented:
> Martin already mentioned this race:

This was due to a last  minute change. Fixing.

petervo commented on test/images/scripts/openshift.setup:
```diff
@@ -8,8 +8,9 @@ xfs_growfs /
+nmcli con up "static-ens2"
+
 printf "OPENSHIFT CONSOLE\n  https://10.111.112.101:8443\n  Login: scruffy  Password: scruffy\n\n" >> /etc/issue
 printf "OPENSHIFT LISTENING ON LOCALHOST\n  $ ssh -NL 8443:localhost:8443 root@10.111.112.101\n\n" >> /etc/issue
```
I don't think these work anymore

stefwalter replied:
They do work if you have a ```cockpit1``` bridge. Or did it not work for you?

petervo replied:
They do, hadn't tried with --network

petervo commented:
I think i got the container tests working. Things seem like they take quite a bit longer to come up but it should be mostly working. However with this method accessing openshift on 10.111.112.101 directly doesn't work, like it did before. So if we stick with this we should remove the xxx-container-shell targets and the ./vm-run kubeconfig hack as those will no longer work.

stefwalter commented:
> However with this method accessing openshift on 10.111.112.101 directly doesn't work, like it did before. So if we stick with this we should remove the xxx-container-shell targets and the ./vm-run kubeconfig hack as those will no longer work.

It does work for users if they run ```vm-run --network``` first. That'll create the ```cockpit1``` bridge, which ```vm-run``` will then use later to allow users to interact with the vms they way they did before. Or did that part not work for you?

petervo commented:
they do, hadn't tried with --network

martinpitt commented:
Retried the tests after fixing the Debian/Ubuntu package builds, now these are "real" test regressions.

stefwalter commented:
RHEL image creation is broken right now. Blocking on #6360

stefwalter commented:
Since we're blocking on RHEL images, I've split out the other image changes, so we can make progress there. In particular updating the Openshift image means a lot of regressions and changed behavior.

petervo commented on test/verify/check-multi-os:
```diff
@@ -56,8 +56,9 @@ def new_add_machine(b, address):
-    additional_machines = {
-        'fedora-stock': { 'machine': { 'image': 'fedora-stock' } }
+    provision = {
+        "0": { "address": "10.111.113.1/20" },
```
Is there a special reason for "0" instead of machine1 like elsewhere?

stefwalter replied:
The order is alphabetical. The first one in the dict becomes self.machine.

stefwalter commented on test/verify/check-docker:
```diff
@@ -25,6 +25,20 @@ import os
+        m.execute("systemctl start docker")
+
+        # HACK: docker fails to start a container without /etc/resolv.conf
+        m.execute("touch /etc/resolv.conf")
```
I'm rebuilding all the images with this and splitting it out into a separate pull request.

martinpitt replied:
Personally I'd much rather leave this hack in the tests (it's cheap to do) rather than "polluting" the images with it, and this is a bug which should eventually be solved in docker proper.

stefwalter replied:
This is unlikely to be solved in Docker proper for a while. Given the age of the bug. But I guess you guys need to decide which way to go here. I've already spent way too many hours on this on doing it two different ways, and at least half of that work (if not all) will be thrown away.

stefwalter commented:
> Also this is a lot slower than before for me. I'm not sure if that's expected or not. It may still be a worthy trade off but I think i'd like to understand why that seems to be case.

@petervo Is this faster now? I think I fixed the problem in #6478

stefwalter commented:
Rebased.

stefwalter commented on test/images/openshift-node:
```diff
@@ -1 +0,0 @@
-openshift
```
Done.

stefwalter commented on test/common/testvm.py:
```diff
@@ -96,25 +91,56 @@ def __str__(self):
+        self.ssh_reachable = False
+
+    def diagnose(self):
+        keys = {
```
I don't think so. We should be getting away from those, and we don't know the URLs for them.

stefwalter commented on test/common/testlib.py:
```diff
@@ -490,50 +519,57 @@ def run(self, result=None):
                 self.currentResult.printErrors()
                 if self.machine:
-                    print >> sys.stderr, "ADDRESS: %s" % self.machine.address
+                    sys.stderr.write(self.machine.diagnose())
```
Done.

martinpitt commented:
The `testBasic (check_kdump.TestKdump)` failure on rhel-7 seems pretty resilient. I'll retry again, but this might be an actual regression.

martinpitt commented:
I ran check-kdump on rhel-7 three times locally, and it always succeeds, so it's not completely broken but just  some timing issue. The test does take rather long (220 s), it spends a lot of time in the last line here:
```py
        # wait for disconnect and then try connecting again
        self.browser.switch_to_top()
        self.browser.wait_present("div.curtains-ct") 
        self.browser.wait_visible("div.curtains-ct")
```

```
-> wait('ph_is_visible("div.curtains-ct")')
> transport closed: disconnected
> transport closed, dropped message:  1:1!4
{"call":["/org/freedesktop/timedate1","org.freedesktop.DBus.Properties","Get",["org.freedesktop.timedate1","NTPSynchronized"]],"id":"18"}
^^^^^^ this line repeats over and over for a minute or longer
<- True
-> wait('ph_in_text("div.curtains-ct h1","Disconnected")')
<- True
```

martinpitt commented:
The `self.rebootMachine()` that the CI is timing out on only takes about 10s locally, so I cannot reproduce that. The `Job for kdump.service failed because the control process exited with error code.` happens locally as well and is just due to `No memory reserved for crash kernel.`, so this is expected.

martinpitt commented:
I pushed PR #6923 to debug this timeout issue.

martinpitt commented:
I think I figured it out. I'll turn #6923 into a "proper" PR (so far it's just been a debugging vehicle) and see to landing that ASAP. Setting this one to "blocked" until then.

martinpitt commented:
#6923 landed now. I force-pushed a rebase to master with dropping the original image rebuilds.  I'll let the bot rebuild ipa and am rebuilding openshift locally.

cockpituous commented:
image-refresh in progress on cockpit-tests-0xj8f.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-6334-20170614-141212/

cockpituous commented:
image-refresh ipa done: https://github.com/cockpituous/cockpit/commits/image-refresh-ipa-20170614-145050

martinpitt commented:
One thing that I noticed is that rebuilding the openshift image locally takes a looooooong time - pulling the docker.io/library/fedora is really slow, it only pulls with 50kB/s. A local docker pull is fast (2 MB/s), so that might be the qemu user networking.

martinpitt commented:
Force-pushed again to fold in the rebuilt ipa and openshift images. Checks will fail right now as my openshift image upload is still going on, I'll retry them once that finished.

martinpitt commented:
This now breaks the new check-packagekit: `'pkcon refresh' returned non-zero exit status 7`. I can't see the output in the non-verbose log, but very likely this is PackageKit complaining about being offline again (which is the change from this PR). It is not wrong, except that the test repos are all local. So I need to find a way to fake it.

I'm AFK until Monday, will deal with that then.

martinpitt commented:
I found a simple way to coerce PackageKit into thinking the system is online. As this also affects CentOS/RHEL-7 now, and the approach is quite generic, I just dropped the OS name check and run that everywhere. I also rebased again to fix the check-multi-machine conflict.

martinpitt commented:
Whoops,  I screwed up the rebase, fixed now (and verified that check-multi-machine succeeds locally).

martinpitt commented:
test/verify/check-multi-machine now fails persistently on rhel-7/centos-7, it didn't before. I can reproduce this locally.

martinpitt commented:
This is a bit ugly - it only seems to happen on rhel/centos-7, and it does not happen when I add a sit() and run the steps interactively, so it is a race condition of some sort. The output doesn't tell us much:
```
Ignoring Resource Error: Operation canceled http://127.0.0.2:9091/=10.111.113.2

Jun 19 02:41:03 localhost.localdomain cockpit-ws[1469]: cockpit-ssh: session timed out during authentication
Jun 19 02:41:03 localhost.localdomain cockpit-ws[1469]: cockpit-ssh: authentication timed out
Jun 19 02:41:03 localhost.localdomain cockpit-ws[1469]: ignoring failure from session process: Authentication failed: Timeout
```
On the remote (machine2) side, the only log entry around that time is
```
Jun 19 02:41:03 machine2 sshd[1632]: Connection closed by 10.111.113.1 [preauth]
```

martinpitt commented:
Debug output on main machine while this happens: not sure what all these "invalid credential cookies" are all about:
```
Jun 19 05:53:39 localhost.localdomain cockpit-ws[1483]: session initialized
Jun 19 05:53:39 localhost.localdomain cockpit-ws[1483]: cockpit-ssh: no-host
Jun 19 05:53:39 localhost.localdomain cockpit-ws[1483]: web service closing
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: spawning /usr/libexec/cockpit-ssh
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received authorize challenge
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received authorize challenge
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: session initialized
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: session is idle
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: logged in user session
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received init message
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received cockpit+=10.111.113.2 credential cookie for session
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: /manifests.json: completed serving external channel
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received cockpit+=10.111.113.2 credential cookie for session
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: /shell/index.html: completed serving external channel
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received unknown/invalid credential cookie
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: spawning /usr/libexec/cockpit-session
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: received authorize challenge
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: session initialized
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: cockpit-session: authentication-unavailable Authentication service cannot retrieve authentication info
Jun 19 05:53:40 localhost.localdomain cockpit-ws[1483]: web service closing
Jun 19 05:53:55 localhost.localdomain cockpit-ws[1483]: session timed out
```

I noticed that "realmd[1231]: quitting realmd service after timeout" roughly coincides with these preauth failures. However, neither running `systemctl stop realmd` nor `start` makes a difference, nor a simple 
`time.sleep(2)`.

stefwalter commented:
Included work from #7060

stefwalter commented:
Rebased.

martinpitt commented:
Uh, for some reason the last rebase has caused a slew of new failures which don't easily go away with retries :-(

martinpitt commented:
Now we are back to the [authentication timeouts](https://github.com/cockpit-project/cockpit/pull/6334#issuecomment-309352530) that we saw earlier. But they happen across all OSes now, I replicated it locally after some 5 attempts on Fedora-26.

stefwalter commented:
Need some feedback on this pull request, so that it can stop rotting. Marking priority.

martinpitt commented:
I'm looking at the timeout of `check-packagekit TestUpdates.testInfoSecurity` on debian-testing. It eventually succeeds to get the changelogs, but only after 1,5 minutes (and the test has a default timeout of 30s).

Apparently apt uses the glibc resolver to get the changelog server address:

```
1609  04:40:33.399316 connect(3, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("172.27.0.3")}, 16) = 0
1609  04:40:53.429203 connect(3, {sa_family=AF_INET, sin_port=htons(12345), sin_addr=inet_addr("localhost")}, 16) = -1 EINPROGRESS (Operation now in progress)
```
Unfortunately using `127.0.0.1` as the changelog server address doesn't help, it will just do the resolution attempt for `127.0.0.1` then.

/etc/resolv.conf has `nameserver 172.27.0.3` which is apparently bogus - ping and traceroute to that fail, and doing e. g. `host localhost` or `dig localhost` time out (`no servers could be reached`). Dropping it from /etc/resolv.conf fixes the long hangs. I'm not entirely sure why that is -- `/etc/hosts` does have localhost, and `/etc/nsswitch.conf` has `hosts: files dns`, and even dropping `dns` doesn't help.

So apparently QEMU's DHCP server sends 172.27.0.3 as a nameserver which isn't working. IMHO the cleanest solution would be to fix it on that side - don't send a nameserver, we don't want tests to resolve non-local names, and we want to fail quickly on those instead of timing out. If that's too hard, we could also hack the test to clear out /etc/resolv.conf or configure the DNS client to ignore/override the sent name server.

stefwalter commented:
Pushed a fix to add DNS support to both the network config, and setting up an empty DNS server when necessary.

martinpitt commented:
FTR: I tried to add `dns=127.0.0.1` to the `-net user` option, but QEMU doesn't like that. It does understand the option (e. g. `dns=172.27.0.5` is accepted), it just seems to refuse IPs that aren't on that network. Stef's approach does seem to work. Thanks!

stefwalter commented:
Rebased.

martinpitt commented:
I pushed a FIXUP which solves the `AttributeError: VirtMachine instance has no attribute 'fetch'` (rebase error). check-machines now succeeds locally. @stefwalter, can you please double-check?

stefwalter commented:
Thanks, squashed.

martinpitt commented:
The previous test run failed `test/verify/check-connection TestConnection.testBasic` with two different expected journal messages. On [centos-7](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170712-052929-6439792f-verify-centos-7/log.html#1) with
```
Unexpected journal message 'g_hash_table_insert_internal: assertion 'hash_table != NULL' failed'
```
and on [debian-testing](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170712-051919-6439792f-verify-debian-testing/log.html#46) with
```
Unexpected journal message 'peer did not close io when expected'
```
I can't reproduce these locally, these look rather "(un)lucky". I think the latter deserves to be added to `allow_restart_journal_messages()` (I sent PR #7206), I'm less sure about the former. My gut feeling is that we leave that as-is, retry here, and see if it happens often.

martinpitt commented:
The `g_hash_table_insert_internal: assertion 'hash_table != NULL' failed` message also happened 
[in a different test on rhel-7-4](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170712-061010-6439792f-verify-rhel-7-4/log.html#25), so that does appear like something which we need to investigate. (Still, not necessarily in this PR).

I can reproduce the two check-storage-raid regressions on debian-testing locally. They happen reliably in this branch (so it'll be rather hard to get them to green here). On master *this* particular failure (failed to stop due to "device or resource busy") does not happen, although `testNotRemovingDisks()` fails there often in a different way. I'm investigating.

martinpitt commented:
check-storage-raid debugging notes:
 * Adding a `time.sleep(5)` before stopping the RAID works; whatever is keeping the RAID busy before is done then.
 * A `ps aux` diff before/after the sleep only shows that `/usr/lib/NetworkManager/nm-dispatcher` went away. However, that appears to  be thoroughly unrelated to the RAID state. Adding a `pkill -9 -e nm-dispatcher` indeed does not help.
 * During these 5 seconds of sleep there is no additional `dmesg` output.

This smells like udev rules probing the device after it gets created, or the mdraid monitor actually reading the device on startup, or something similar. I propose we just accept that there is a small period of time after creating a RAID after which the device is busy, and retry up to three times, with handling the "busy" error dialog.

martinpitt commented:
Force-pushed a rebase (simple conflicts from the debian-stable update) and added a first fix for the RAID failures. This fixes `TestStorage.testRaid`, but `TestStorage.testNotRemovingDisks` still fails for a different reason later on.

martinpitt commented:
The remaining RAID failure is
```
$ test/verify/check-storage-mdraid TestStorage.testNotRemovingDisks
  File "test/verify/check-storage-mdraid", line 306, in testNotRemovingDisks
    self.wait_states({ "QEMU QEMU HARDDISK (DISK1)": "FAILED" })
Error: timeout
```
The UI screenshot indeed still shows the first hard disk status as "in sync", but when I log in, I see it as "FAILED", plus a big red warning "The RAID Array is in a degraded state" on top.

That indeed seems to be a failure of underlying udisks2: I was running `udisksctl monitor` during the test, and when it runs `mdadm --fail`, there is no property change from udisks. Only *after* the test times out, the monitor finally shows
```
05:22:00.970: /org/freedesktop/UDisks2/mdraid/ecbabc7f_cd160994_94c0cf91_e0f70442: org.freedesktop.UDisks2.MDRaid: Properties Changed
  ActiveDevices:        [('/org/freedesktop/UDisks2/block_devices/sda', -1, ['faulty'], 0, {}), ('/org/freedesktop/UDisks2/block_devices/sdb', 1, ['in_sync'], 0, {})]
  Degraded:             1
```

Curiously, if I log into Cockpit from the browser (http://127.0.0.2:9091), there is no delay, udisks sends the property change immediately, and the test succeeds. So it feels like phantom doesn't run the main loop hard enough in `wait_states()`? I tried bumping the timeout to two minutes, doesn't help. (Maybe because `wait_states()` doesn't call any phantom function?)

martinpitt commented:
Force-pushed again to fix [this test failure](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170712-100707-83a516bf-verify-debian-stable/log.html#6), collateral damage from moving debian-stable to 9.

martinpitt commented:
The [network tests on rhel-7.4](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170712-124141-3885c5b9-verify-rhel-7-4/log.html) keep failing with a very weird error (`curl: (23) Failed writing body (265 != 16384)`) when repeatedly trying to download a Fedora-26 image (!?). There's something seriously wrong there. I retry again for now.

martinpitt commented:
The above was not just a flake, it just [happened again on ubuntu-1604](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170712-140707-3885c5b9-verify-ubuntu-1604/log.html#49).  For some reason the various check-networking-* tests attempt to download a fedora-26 image, on *any* OS. I believe that should be fixed.

martinpitt commented:
I stand corrected - `check-networking-*` actually *do* have a `provision` which adds a fedora-26. (Sorry, I was checking in master before.) But I wonder why? It seems these get added in vain, as the actual tests don't make use of them.

stefwalter commented:
> But I wonder why? It seems these get added in vain, as the actual tests don't make use of them.

They need a DHCP server, and use fedora-26 image along with provision and a dhcp argument to set one up.

martinpitt commented:
The image downloads are killing us here. I filed PR #7236 to hopefully rectify this, although admittedly this is a bit monkey-patching.

stefwalter commented:
Another change here:

```
diff --git a/test/common/cloud-init.iso b/test/common/cloud-init.iso
index 5573e533d..3ffbb0828 100644
Binary files a/test/common/cloud-init.iso and b/test/common/cloud-init.iso differ
diff --git a/test/common/make-cloud-init-iso b/test/common/make-cloud-init-iso
index 336cf6962..d4b6d80da 100755
--- a/test/common/make-cloud-init-iso
+++ b/test/common/make-cloud-init-iso
@@ -75,10 +75,6 @@ cloud_final_modules:
  - [scripts-user, always]
  - final-message
 
-# we may need to ping something on the network to update the arp cache
-runcmd:
-  - while ! ifconfig | grep -q 10\.111\.; do sleep .1; done; ping -W 1s -c 3 10.111.112.1 || true
-
 EOF
 
 genisoimage -input-charset utf-8 -output $iso_image -volid cidata -joliet -rock $user_data $meta_data
```

martinpitt commented:
Force-pushed again to (re-)fix [testTls](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170713-063131-5a8e7852-verify-debian-stable/log.html#6). That got lost in the previous rebase. @stefwalter , can you please reset your local checkout?

stefwalter commented:
The rhel-atomic and fedora-atomic images need to be regenerated to remove the above runcmd script which gets placed in them during their first boot.

stefwalter commented:
Switched to ```systemctl reload NetworkManager``` instead of ```systemctl restart NetworkManager``` after setting up resolv.conf. The latter caused ubuntu-stable systemctl command to just hang forever.

martinpitt commented:
The `testNegotiate (check_realms.TestKerberos)` [test](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170713-200808-813b18d7-verify-ubuntu-stable/log.html#148) consistently fails on ubuntu, and takes 40 minutes to time out. 

The ssh  fails because it cannot resolve its own new hostname (`x0.cockpit.lan`):
```
+ ssh -oStrictHostKeyChecking=no -oBatchMode=yes -l admin@cockpit.lan x0.cockpit.lan true
ssh_exchange_identification: Connection closed by remote host
```
```
# host x0.cockpit.lan
Host x0.cockpit.lan not found: 3(NXDOMAIN
```

Ubuntu does not install `nss-myhostname` by default. Fedora does, so `hostnamectl set-hostname` suffices there. Once I add that to `/etc/hosts`, it can connect, but fails because the key is not authenticated:
```
+ ssh -oStrictHostKeyChecking=no -oBatchMode=yes -l admin@cockpit.lan x0.cockpit.lan true
Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password,keyboard-interactive).
```

I just realized that resolving it to `127.0.0.1` is most probably wrong, it should be `10.111.113.1`. Unfortunately the test now keeps failing (quickly) on `kinit: Cannot contact any KDC for realm 'COCKPIT.LAN'`, not sure what changed (there is no leftover VM or process on my machine, and this should all happen in the VM anyway). I feel a bit out of luck today, but I'll continue to investigate tomorrow - this is most likely a simple fix, like
```patch
--- a/test/verify/check-realms
+++ b/test/verify/check-realms
@@ -207,6 +207,9 @@ class TestKerberos(MachineCase):
         # Setup a place for kerberos caches
         args = { "addr": "10.111.112.100", "password": "foobarfoo" }
         self.machine.execute("hostnamectl set-hostname x0.cockpit.lan")
+        if "ubuntu" in self.machine.image:
+            # no nss-myhostname there
+            self.machine.execute("echo '10.111.113.1 x0.cockpit.lan' >> /etc/hosts")
         self.machine.execute(script=JOIN_SCRIPT % args)
 
     def tearDown(self):
```

If that were the only remaining failure I'd say we skip it for now and investigate/fix in a separate PR. But apparently it's not - `check-ostree` on rhel-atomic also [doesn't look well](https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170713-204949-813b18d7-verify-rhel-atomic/log.html#4), and various networking tests across the board are really hard to get right (but that might again just be because they fail more reliably now, not necessarily because this PR makes things worse).

martinpitt commented:
... and that now actually worked - I pushed this as a fixup. However, I just realized that my rationale in the commit was bogus though, as `RESOLV_SCRIPT` did not actually add the host name to `/etc/hosts`, but to `resolv.conf` as a name server. Presumably there is something on the localhost that should act as a name server, but at least on Fedora 26 this is done by `nss-myhostname`. So adding `x0.cockpit.lan` to `/etc/hosts` on Ubuntu has at least an equivalent effect, even though I'm not sure if the test is fully *intended* to work like this.

But my gut feeling is that it's good enough to make progress on this PR, unless you object (easy enough to revert).

stefwalter commented:
Squashed the fixup.

> But apparently it's not - check-ostree on rhel-atomic also doesn't look well,

This is most likely due to the fact that the rhel-atomic image was regenerated. I wanted to avoid that, but the runcmd was placed there via cloud-init on first boot, and then stuck around in the image.

> and various networking tests across the board are really hard to get right (but that might again just be because they fail more reliably now, not necessarily because this PR makes things worse).

stefwalter commented:
Lets do a follow up for: https://fedorapeople.org/groups/cockpit/logs/pull-6334-20170714-123131-d7bde0d0-verify-fedora-atomic/log.html#79
